compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1660,10 +1660,19 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         // it's a actual definition. According to the comments (e.g. in
         // librustc_typeck/check/compare_method.rs:compare_predicate_entailment) the latter
         // is relied upon by some other code. This might (or might not) need cleanup.
-        let body_owner_def_id =
+        let mut body_owner_def_id =
             self.tcx.hir().opt_local_def_id(cause.body_id).unwrap_or_else(|| {
                 self.tcx.hir().body_owner_def_id(hir::BodyId { hir_id: cause.body_id })
             });
+
+        if let hir::def::DefKind::AnonConst = self.tcx.hir().def_kind(body_owner_def_id) {
```

Comment:
This seems quite subtle, for instance in being sure that `AnonConst` is the only special case. Are there no general methods for doing this (and should there be, if this pattern is used elsewhere)?

Reply:
IIUC a `DefId` can be given to any of the variants of [`DefKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/def/enum.DefKind.html)? Maybe closures/generators can cause problems too?